x-pack/legacy/plugins/apm/server/lib/alerting/error_occurrence/create_alert.ts

```diff
@@ -0,0 +1,158 @@
+}
+
+interface Action {
+  actionTypeId: string;
+  description: string;
+  config: Record<string, any>;
+  secrets: Record<string, any>;
+}
+const createAlert = async (
+  { alertsClient, actionsClient }: Clients,
+  { actions: { slack, email }, serviceName, threshold, interval }: Properties
+) => {
+  const actions: Array<{ params: Record<string, any>; action: Action }> = [];
+
+  const values = {
+    errorLogMessage: '{{{context.errorLogMessage}}}',
+    errorCulprit: '{{{context.errorCulprit}}}',
+    docCount: '{{context.docCount}}',
+    serviceName,
+    threshold,
+    interval
+  };
+
+  if (email) {
+    actions.push({
+      action: {
+        actionTypeId: '.email',
+        description: 'Email notifications for error occurrences',
+        config: {
+          from: ''
+        },
+        secrets: {
+          user: '',
+          password: ''
+        }
+      },
+      params: {
+        to: email,
+        subject: i18n.translate(
+          'xpack.apm.serviceDetails.enableErrorReportsPanel.emailSubjectText',
+          {
+            defaultMessage:
+              '{serviceName} has error groups which exceeds the threshold',
+            values: { serviceName }
+          }
+        ),
+        body: i18n.translate(
+          'xpack.apm.serviceDetails.enableErrorReportsPanel.emailTemplateText',
+          {
+            defaultMessage:
+              `Your service {serviceName} has error groups which exceeds {threshold} occurrences within {interval}<br/>` +
+              '<br/>' +
+              '{errorLogMessage}<br/>' +
+              '{errorCulprit}<br/>' +
+              '{docCount} occurrences<br/>',
+            values
+          }
+        )
+      }
+    });
+  }
+
+  if (slack) {
+    actions.push({
+      action: {
+        actionTypeId: '.slack',
+        description: 'Slack notifications for error occurrences',
+        config: {},
+        secrets: {
+          webhookUrl: slack
+        }
+      },
+      params: {
+        message: i18n.translate(
+          'xpack.apm.serviceDetails.enableErrorReportsPanel.slackTemplateText',
+          {
+            defaultMessage: `Your service *{serviceName}* has error groups which exceeds {threshold} occurrences within {interval}
+>*{errorLogMessage}*
+>\`{errorCulprit}\`
+>{docCount} occurrences`,
+            values
+          }
+        )
+      }
+    });
+  }
+
+  if (!actions.length) {
+    throw new Error(
+      'No actions were defined. Need at least one of email, slack'
+    );
+  }
+
+  const actionResults = await Promise.all(
+    actions.map(({ action, params }) => {
+      return actionsClient
+        .create({
+          action: {
+            actionTypeId: action.actionTypeId,
+            description: action.description,
+            config: action.config,
+            secrets: action.secrets
+          }
+        })
+        .then(result => ({
+          result,
+          params
+        }));
+    })
+  );
+
+  return alertsClient.create({
```

Comment:
Not sure if we're allowed to do this or if we have to use the API.

Reply:
This is ok to do, all the business logic currently lives within the client and you're getting the client from the request (which is good). You will just have to handle validation until we have some within the client (if we decided to do so).